RDoc-2630 [Python] Client API > Session > Open a session [Replace C# samples] (#1771)
ravendb/docs: Documentation/5.4/Raven.Documentation.Pages/client-api/session/opening-a-session.python.markdown

# Open a Session

---

{NOTE: }

* A Session object is obtained from the [Document Store](../../client-api/what-is-a-document-store).  

* Open a session using `open_session()`.  

* Various Session options can be configured using the `SessionOptions` object.  
  If no database is specified in the options then the [Default Database](../../client-api/setting-up-default-database) (stored in the Document Store) is assumed.  

> * Be sure to wrap the session variable with a 'using' statement to ensure proper disposal.

Reviewer: 'with' statement

Author: done

> * Be sure to wrap the session variable with a 'using' statement to ensure proper disposal.

Reviewer: ```suggestion
* Be sure to wrap the session variable using a 'with' statement to ensure proper disposal.
```

Author: done

* In this page:  
  * [Syntax](../../client-api/session/opening-a-session#syntax)  
  * [Session options](../../client-api/session/opening-a-session#session-options)  
  * [Open session example](../../client-api/session/opening-a-session#open-session-example)
{NOTE/}

---

{PANEL:Syntax}

* Use `open_session()` to open a session from the Document Store.  

> * The following overloads are available:  

Reviewer: i think we should drop this line, we have no overloads here

Author: done

| Return Value                                 | Description                   |
|----------------------------------------------|-------------------------------|
| `IDocumentSession` | Instance of a Session object  |

{PANEL/}

{PANEL:Session options}

* The `SessionOptions` object contains various options to configure the Session's behavior.

| Option                                                  | Type              | Description                                                                                                                                                                                                                               | Default Value                                         |
|---------------------------------------------------------|-------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------|

> | **Database**                                            | string            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `null` - the Session operates on the Default Database |

Reviewer: `snake_case` every option

> | **Database**                                            | string            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `null` - the Session operates on the Default Database |

Reviewer: ```suggestion
| **Database**                                            | string            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `None` - the Session operates on the Default Database |
```

in Python instead of null we use `None`

Author: done

> | **Database**                                            | string            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `null` - the Session operates on the Default Database |

Reviewer: ```suggestion
| **database**                                            | string            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `null` - the Session operates on the Default Database |
```

Author: done

> | **Database**                                            | string            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `null` - the Session operates on the Default Database |

Reviewer: ```suggestion
| **Database**                                            | str            | The Session will operate on this database,<br>overriding the Default Database.                                                                                                                                                            | `null` - the Session operates on the Default Database |
```

Reviewer: use `str` instead of 'string'

Author: done

> | **NoTracking**                                          | bool              | `true` - The Session tracks changes made to all entities it loaded, stored, or queried for.<br>`false` - Tracking will be turned off.<br>Learn more in [Disable tracking](../../client-api/session/configuration/how-to-disable-tracking) | `false`                                               |

Reviewer: ```suggestion
| **NoTracking**                                          | bool              | `True` - The Session tracks changes made to all entities it loaded, stored, or queried for.<br>`False` - Tracking will be turned off.<br>Learn more in [Disable tracking](../../client-api/session/configuration/how-to-disable-tracking) | `false`                                               |
```

In Python True and False begin with capital letter

Author: done

> | **NoCaching**                                           | bool              | `true` - Server responses will Not be cached.<br>`false` - The Session caches the server responses.<br>Learn more in [Disable caching](../../client-api/session/configuration/how-to-disable-caching)                                     | `false`                                               |

Reviewer: ```suggestion
| **NoCaching**                                           | bool              | `true` - Server responses will Not be cached.<br>`false` - The Session caches the server responses.<br>Learn more in [Disable caching](../../client-api/session/configuration/how-to-disable-caching)                                     | `false`                                               |
```
```suggestion
| **no_caching**                                           | bool              | `true` - Server responses will Not be cached.<br>`false` - The Session caches the server responses.<br>Learn more in [Disable caching](../../client-api/session/configuration/how-to-disable-caching)                                     | `false`                                               |
```

> | **no_caching**                                          | bool               | `True` - Server responses will Not be cached.<br>`False` - The Session caches the server responses.<br>Learn more in [Disable caching](../../client-api/session/configuration/how-to-disable-caching)                                     | `False`                                               |

Reviewer: ```suggestion
| **no_caching**                                          | bool               | `True` - Server responses will not be cached.<br>`False` - The Session caches the server responses.<br>Learn more in [Disable caching](../../client-api/session/configuration/how-to-disable-caching)                                     | `False`                                               |
```

Author: done

> | **request_executor**                                    | `request_executor` | ( _Advanced option_ ) <br>The request executor the Session should use.                                                                                                                                                                    | `None` - the default request executor is used         |

Reviewer: ```suggestion
| **request_executor**                                    | `RequestExecutor` | ( _Advanced option_ ) <br>The request executor the Session should use.                                                                                                                                                                    | `None` - the default request executor is used         |
```

Author: done

> | **TransactionMode**                                     | `TransactionMode` | Specify the Session's transaction mode<br>`SingleNode` / `ClusterWide`<br>Learn more in [Cluster-wide vs. Single-node](../../client-api/session/cluster-transaction/overview#cluster-wide-transaction-vs.-single-node-transaction)        | `SingleNode`                                          |

Reviewer: ```suggestion
| **TransactionMode**                                     | `TransactionMode` | Specify the Session's transaction mode<br>`SINGLE_NODE / `CLUSTER_WIDE`<br>Learn more in [Cluster-wide vs. Single-node](../../client-api/session/cluster-transaction/overview#cluster-wide-transaction-vs.-single-node-transaction)        | `SingleNode`                                          |
```

Author: done

Author: also changed the default value in the last column (--> SINGLE_NODE)

> | **transaction_mode**                                    | `transaction_mode` | Specify the Session's transaction mode<br>`SINGLE_NODE` / `CLUSTER_WIDE`<br>Learn more in [Cluster-wide vs. Single-node](../../client-api/session/cluster-transaction/overview#cluster-wide-transaction-vs.-single-node-transaction)      | `SINGLE_NODE`                                         |

Reviewer: ```suggestion
| **transaction_mode**                                    | `TransactionMode` | Specify the Session's transaction mode<br>`SINGLE_NODE` / `CLUSTER_WIDE`<br>Learn more in [Cluster-wide vs. Single-node](../../client-api/session/cluster-transaction/overview#cluster-wide-transaction-vs.-single-node-transaction)      | `SINGLE_NODE`                                         |
```

Author: done

* Experts Only:

| Option                                                       | Type                | Description                                                                                                                                                                                                                                             | Default Value |
|--------------------------------------------------------------|---------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------|

> | **DisableAtomicDocumentWrites-<br>InClusterWideTransaction** | bool?               | ( _Experts only_ ) <br>`true` - Disable Atomic-Guards in cluster-wide sessions.<br>`false` - Automatic atomic writes in cluster-wide sessions are enabled.<br>Learn more in [Atomic-Guards](../../client-api/session/cluster-transaction/atomic-guards) | `false`       |

Reviewer: there's no `?` operator in Python like in c#
```suggestion
| **DisableAtomicDocumentWrites-<br>InClusterWideTransaction** | bool               | ( _Experts only_ ) <br>`true` - Disable Atomic-Guards in cluster-wide sessions.<br>`false` - Automatic atomic writes in cluster-wide sessions are enabled.<br>Learn more in [Atomic-Guards](../../client-api/session/cluster-transaction/atomic-guards) | `false`       |
```

Author: done
